### ISSUE TITLE
[feature] Conan simplify build procedure

### ISSUE THREAD

**Issue by redradist** [conan-io/conan#6835]
**C and C++** known to be hard to work with, because of lots of reasons ...
One of the reason that user should know how to install particular toolchain and configure it

Such languages as **Rust** has **RustUp**, but we do not (:

What if Conan will add downloading, configuring and caching of toolchain ?!

Consider use-case:

```bash
conan download --toolchain clang-arm
conan build
```

It will simplify working with C and C++ ;)

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Conan can install tools and compilers wrapped in conan packages

The following works in Windows:
```
$ conan install mingw_installer/1.0@conan/stable -g virtualrunenv -s compiler=gcc -s compiler.exception=seh -s compiler.threads=win32 -s compiler.version=6.3
$ activate_run.bat
$ gcc --version
```

There are several reasons why Conan does not provide built-in toolchains:
- There are license issues, Conan cannot wrap and redistribute certain toolchains, especially popular ones like Visual Studio, without licensing issues
- Many toolchains cannot easily be packaged and installed side by side, but they are kind of system-wide, and require tools as ``update-alternatives``
- A vast majority of Conan users already have their toolchains already installed by other means (IT, docker, etc)
- The users that use the above feature typically have their own very customized compilers or toolchain versions. In C and C++ is difficult to have a one size fits all

So as a summary, this would be a really big effort (packaging the toolchains, storage and bandwidth, maintenance and making sure they work in a variety of platforms, distros, etc), for a marginal value for the overall Conan community of users. The ecosystem is wildly different from the one of Rust (starting from the huge team of maintainers and core contributors that Rust has behind, compared with the Conan one), and the same principles do not apply.

If you want to see this happens, the best is to start contributing packages containing toolchains to the ConanCenter, the central repository of Conan packages: https://github.com/conan-io/conan-center-index. It the packages there get enough community support and are accepted, that would be  a good start, and in the future they could be managed more like built-in things. Please file issues first to discuss and get community feedback, interest, potential contributors. Thanks!

As this cannot be a Conan feature at the moment, closing this, please follow conversation in https://github.com/conan-io/conan-center-index if you wish. Thanks for the feedback!

**Comment by redradist** [comment]
@memsharded 
> There are several reasons why Conan does not provide built-in toolchains:
> 
>     * There are license issues, Conan cannot wrap and redistribute certain toolchains, especially popular ones like Visual Studio, without licensing issues

Then Visual Studio will not be available through the conan

>     * Many toolchains cannot easily be packaged and installed side by side, but they are kind of system-wide, and require tools as `update-alternatives`

Not really true, for example you can provide to your `cmake` project external toolchain described in special file
There are examples:
https://github.com/ruslo/polly

>     * A vast majority of Conan users already have their toolchains already installed by other means (IT, docker, etc)
It is because they do not have centralize management for toolchains like Rust projects have with `rustup`

>     * The users that use the above feature typically have their own very customized compilers or toolchain versions. In C and C++ is difficult to have a one size fits all

Not really true, I worked on many projects and most of the time it just copy of toolchain from Linario, ESP32, Clang and other resources

> So as a summary, this would be a really big effort (packaging the toolchains, storage and bandwidth, maintenance and making sure they work in a variety of platforms, distros, etc), for a marginal value for the overall Conan community of users. The ecosystem is wildly different from the one of Rust (starting from the huge team of maintainers and core contributors that Rust has behind, compared with the Conan one), and the same principles do not apply.

Not really, again ;)
With your conanfile.py you already have all needed infrastructure for packaging toolchain ...
What you do not have:
1) Special flags for package that it is toolchain
2) Generation toolchain file (can be done by generating from profile file)

> If you want to see this happens, the best is to start contributing packages containing toolchains to the ConanCenter, the central repository of Conan packages: https://github.com/conan-io/conan-center-index. It the packages there get enough community support and are accepted, that would be a good start, and in the future they could be managed more like built-in things. Please file issues first to discuss and get community feedback, interest, potential contributors. Thanks!
> 
> As this cannot be a Conan feature at the moment, closing this, please follow conversation in https://github.com/conan-io/conan-center-index if you wish. Thanks for the feedback!

Started conversation there ;)

**Comment by memsharded** [comment]
> Not really true, for example you can provide to your cmake project external toolchain described in special file
There are examples:
https://github.com/ruslo/polly

That is not distributing the Microsoft build toolchain, it is distributing a cmake toolchain file to use it, but that still needs the user to manually download and install the Microsoft build tools manually (and accept their license, etc). That is one of the main reasons, it is still not possible to package the Microsoft toolchain and automate its installation, I really wish it was possible, but so far it seems it is against their license.

So for Microsoft, so far, this is not possible at all, it cannot be automated Rustup style, until Microsoft changes their license.

> Not really true, I worked on many projects and most of the time it just copy of toolchain from Linario, ESP32, Clang and other resources

@uilianries has been building different flavors of compilers and std libraries for some Linux distributions. He can give some feedback about the effort that is building compilers binaries from source (which is part of the effort that needs to be done to package them for Conan). It is a very large effort compared to install the same compilers with ``apt-get`` or similar.

Other users have been creating packages for some of the Clang toolchain, and it was also challenging.


By the way, there are new build system integrations now, like ``CMakeToolchain`` that generates ``conan_toolchain.cmake`` files, for integration, but that still do not install the compilers themselves.

> Started conversation there ;)

Great, if you think this is possible, and can contribute packages for compilers like gcc or clang that would simplify the usage of such compilers, that would be fantastic, a very welcomed addition, lets see the discussion in conan-center-index.

**Comment by redradist** [comment]
@memsharded According the feed back from community in conan-center-index https://github.com/conan-io/conan-center-index/issues/6149#issuecomment-877837283 we need to discuss it here

**Comment by memsharded** [comment]
Hi @redradist 

I think there were some misunderstanding. I am saying that this doesn't need to be a client feature, because it is already possible to use toolchains, compilers, etc. inside Conan packages with the Conan client. We don't need to implement this feature, it is already implemented. Have a look to the https://conan.io/center/mingw-w64 package, that can already be used to use Mingw in Windows. 

All that is necessary is to create those packages containing the tools, and the right place to create them is to add them in ConanCenter. There are already some other tools and compilers there, and it is possible to add more.

**Comment by redradist** [comment]
@memsharded Gatcha, but I am talking about managing toolchain with ability to generate toolchain file for cmake
You are taking about building package with external compiler ...

But how to find this compiler ... to build with it ... there more than just `build_requires` option in package needed

**Comment by memsharded** [comment]
Same: it is already possible, and it is already being done in ConanCenter: https://conan.io/center/ios-cmake. 
The package can build and package cmake files with the toolchain, and make it available to consumers of the package with ``self.env_info.CONAN_CMAKE_TOOLCHAIN_FILE = tool_chain`` for example.

**Comment by redradist** [comment]
@memsharded It is not the same
I am taking from user point of view, and you saying that it is possible ...
Even if it is possible, it is not convenient for user to manage such special packages

All this packages are special, it would be nice to mark them in special way and handle little bit differently

Even `dotnet` has possibility to install some external tools: `dotnet tool`

Conan could manager separately toolchains, allow to view all installed toolchains, remove if needed and so on

@madebr @prince-chrismc @SSE4 What do you think ?

**Comment by madebr** [comment]
@redradist 
Do you want a feature to get a get a toolchain through conan?
e.g. fetch mingw automatically + create a profile for it?
```
$ conan profile new mingw --toolchain=mingw-w64/8.1@
Downloading mingw-w64/8.1@
...
Creating profile `mingw`
...
$ conan profile show mingw
onfiguration for profile mingw:

[settings]
os=Windows
arch=x86_64
compiler=gcc
compiler.version=8
compiler.libcxx=libstdc++11
compiler.threads=win32
compiler.exception=seh
build_type=Release
[options]
[conf]
[build_requires]
mingw-w64/8.1@
[env]
AR=x86_64-w64-mingw32-ar
AS=x86_64-w64-mingw32-as
CC=x86_64-w64-mingw32-gcc
CXX=x86_64-w64-mingw32-g++
CPP=x86_64-w64-mingw32-cpp
LD=x86_64-w64-mingw32-ld
NM=x86_64-w64-mingw32-nm
OBJCOPY=x86_64-w64-mingw32-objcopy
OBJDUMP=x86_64-w64-mingw32-objdump
RANLIB=x86_64-w64-mingw32-ranlib
RC=x86_64-w64-mingw32-windres
STRIP=x86_64-w64-mingw32-strip
$ conan create . some_recipe/3.14@ -pr mingw
...
```
This would make it possible to start building without first needing to install a toolchain.

**Comment by prince-chrismc** [comment]
I think that would be a solution...

I recently worked with other languages and ecosystems... 
Not much with rust but in Golang cross compiling or "Targeting an build run environment" is incredibly easy.

You do not need to download and configure tools, re do you environment variables , etc...

You simply change `GOOS` and `GOARC`. You can also run `go tool dist` to see the support platforms

***

The problem is C++ developer, we have so many tools and so many platforms and the ISO wont standardize tooling so everything is complex

Conan can do this... it's just complicated like everything else in C++ :)

**Comment by redradist** [comment]
> 
> 
> @redradist
> Do you want a feature to get a get a toolchain through conan?
> e.g. fetch mingw automatically + create a profile for it?
> 
> ```
> $ conan profile new mingw --toolchain=mingw-w64/8.1@
> Downloading mingw-w64/8.1@
> ...
> Creating profile `mingw`
> ...
> $ conan profile show mingw
> onfiguration for profile mingw:
> 
> [settings]
> os=Windows
> arch=x86_64
> compiler=gcc
> compiler.version=8
> compiler.libcxx=libstdc++11
> compiler.threads=win32
> compiler.exception=seh
> build_type=Release
> [options]
> [conf]
> [build_requires]
> mingw-w64/8.1@
> [env]
> AR=x86_64-w64-mingw32-ar
> AS=x86_64-w64-mingw32-as
> CC=x86_64-w64-mingw32-gcc
> CXX=x86_64-w64-mingw32-g++
> CPP=x86_64-w64-mingw32-cpp
> LD=x86_64-w64-mingw32-ld
> NM=x86_64-w64-mingw32-nm
> OBJCOPY=x86_64-w64-mingw32-objcopy
> OBJDUMP=x86_64-w64-mingw32-objdump
> RANLIB=x86_64-w64-mingw32-ranlib
> RC=x86_64-w64-mingw32-windres
> STRIP=x86_64-w64-mingw32-strip
> $ conan create . some_recipe/3.14@ -pr mingw
> ...
> ```
> 
> This would make it possible to start building without first needing to install a toolchain.

@madebr Yes, I want something like you described ...
This would be very similar how [`rustup`](https://rustup.rs) works

Single CLI to:
- Manage toolchains: install, remove
- Along with installation of toolchain create profile file
- ...

Currently in C++ community it is hard to get a new toolchain on platform.
You need to now url, also some toolchains needed to be built and so on ... all this steps make managing toolchains painful

In contrast, Rust community has `rustup` that allow to manager all toolchains
In C++ world this role could be performed by conan

**Comment by redradist** [comment]
> The problem is C++ developer, we have so many tools and so many platforms and the ISO wont standardize tooling so everything is complex
> 
> Conan can do this... it's just complicated like everything else in C++ :)

@prince-chrismc But conan center index already has some toolchain that are built
They are marked as `build_requires`

We just need to provide in conan receipt special flag that will mark this package as toolchain and also options that described the toolchain

Then we need support in conan for such special package/tools

It is not so hard if just to start working on it

**Comment by prince-chrismc** [comment]
I wholeheartedly agree, we have almost everything it takes. I would even go so far as to say Conan _must_ do this to become the "C++ Package Manager" since it would make a single entry point for everything in the ecosystem (except installing MSVC licenses) which the closest has been CMake but that never made it.

> It is not so hard if just to start working on it

But you need to understand the burden of maintaining such feature grows exponentially... In Rust and Golang... there's only one compiler and the only other variable is the C runtime both have standardized the build tool... The only difference in the packes is the architecture. In Golang there's no notion of build option which alters the binary to my knowledge.

C++ does not have that consistency which makes the job here for Conan difficult to even comprehend.

Despite desperately needing this feature... I am with James, this (in today's C++ ecosystem) should not be solved within the client. Most C++ developers are very happy picking and setting up each tool and manually configuring the environment.
For v2 I think offering toolchain from ConanCenter is an excellent start, to prove this is feasible, then finding consumer demand?
Perhaps this could become a v3 target?

**Comment by redradist** [comment]
@prince-chrismc
> 
> 
> I wholeheartedly agree, we have almost everything it takes. I would even go so far as to say Conan _must_ do this to become the "C++ Package Manager" since it would make a single entry point for everything in the ecosystem (except installing MSVC licenses) which the closest has been CMake but that never made it.

I completely agree with this

> 
> > It is not so hard if just to start working on it
> 
> But you need to understand the burden of maintaining such feature grows exponentially... In Rust and Golang... there's only one compiler and the only other variable is the C runtime both have standardized the build tool... The only difference in the packages is the architecture. In Golang there's no notion of build option which alters the binary to my knowledge.

Not really true.
There are only several toolchains that needed to be supported at first MVP prototype:
- MSVC
- gcc
- clang + target architectures
- apple-clang

Second step would be to communicate with main toolchain stack-holders and to ask to support this in future, also discuss what's features are missed

> 
> C++ does not have that consistency which makes the job here for Conan difficult to even comprehend.
> 
> Despite desperately needing this feature... I am with James, this (in today's C++ ecosystem) should not be solved within the client. Most C++ developers are very happy picking and setting up each tool and manually configuring the environment.
> For v2 I think offering toolchain from ConanCenter is an excellent start, to prove this is feasible, then finding consumer demand?
> Perhaps this could become a v3 target?

It is contradictory statement ... First you agree with me, then you ask to wait to conan v3 ?!
C++ community have been waiting for such tool like conan for years, it is enough waiting ...

From my experience, what stops people from making some innovative stuff is fear that it will be hard to maintain, it could be not feature complete, nobody will use it and so on
I think in such way: If something is very important you need to at least try to do it, even if it will fail

For MVP it should not be hard to implement at least gcc and clang + target architectures

I could try to do this but I do not want spend my time if conan repo will not accept this work afterall
The work could be split between developers on support gcc and clang + target architectures platforms, one thing we need agree the interface for installing toolchain and generating profile, also how package will be marked as toolchain package

**Comment by memsharded** [comment]
Marking them as "toolchain" packages doesn't provide any immediate advantage, besides possibly some search capabilities. But the usage pattern is already well defined, it can be used via ``[build_requires]`` in the profile. Conan 2.0 has already defined a PackageType that can be used to define the type of the package, but still the consumption pattern will be driven by ``build_requires``.

At the moment we are not going to add a special feature to define a profile from a Conan package, and this will require a little bit of repetition. Not a big issue, the 95% of the value of this feature is having the toolchain packaged and ready to be used, and this can already be implemented with Conan packages, as @madebr clearly pointed with the ``mingw`` package, that is already working. The profile in that example can be greatly simplified, because the whole ``[env]`` section doesn't need to be declared in the profile, in can be propagated directly from the package recipe.

And we are not adding more features at the moment, until the compiler packages are there for several reasons:
- Building those packages is a challenge. Building clang is challenging, it requires lots of memory and resources, a lot of time, etc. There have been already a few brilliant people working on this. It is better to know first how far can we get.
- It is impossible to implement this for two thirds of the ecosystem. Licenses do not permit to repackage Microsoft or Apple toolchains, so whatever could be done here would only reach one third of our users. Comparing with go or rust is completely invalid in our ecosystem, I would also love the same experience in C++, but it is simply not realistic.

So packages for Gcc and Clang for Linux will be accepted in ConanCenter, as long as they pass the requirements and the community approves them. In Conan 2.0 their package type will be the adequate one (but still be used as "build_requires"). Once they are there, and there is evidence of usage by the community, we can get back to this issue and analyze if there are other possible UX improvements to simplify using them.

**Comment by redradist** [comment]
@memsharded 
> 
> 
> Marking them as "toolchain" packages doesn't provide any immediate advantage, besides possibly some search capabilities. But the usage pattern is already well defined, it can be used via `[build_requires]` in the profile. Conan 2.0 has already defined a PackageType that can be used to define the type of the package, but still the consumption pattern will be driven by `build_requires`.

It is not enought ...
I do not want to have a copy of `llvm` toolchain, I want to have one toolchain and use it with or without conan packaging stuffs. It is much broader that a simple search of toolchain: search, manager (save, delete), create profile for it and so on

> 
> At the moment we are not going to add a special feature to define a profile from a Conan package, and this will require a little bit of repetition. Not a big issue, the 95% of the value of this feature is having the toolchain packaged and ready to be used, and this can already be implemented with Conan packages, as @madebr clearly pointed with the `mingw` package, that is already working. The profile in that example can be greatly simplified, because the whole `[env]` section doesn't need to be declared in the profile, in can be propagated directly from the package recipe.

How to create profile for `mingw` package ?
How to figure out that it is toolchain ?

**Comment by memsharded** [comment]
> I do not want to have a copy of llvm toolchain, I want to have one toolchain and use it with or without conan packaging stuffs. It is much broader that a simple search of toolchain: search, manager (save, delete), create profile for it and so on

```bash
conan install mingw-w64/8.1@ -g virtualenv
activate.bat
gcc --version
gcc (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
```
And done, it works. You would have the compiler environment defined, you could call ``gcc`` directly in the command line, no Conan involved.

> How to create profile for mingw package ?

Write a file with:
```
[settings]
arch=x86_64
os=Windows
compiler=gcc
compiler.version=8.1
compiler.libcxx=libstdc++11
compiler.threads=posix
compiler.exception=seh

[build_requires]
mingw-w64/8.1
```

This can be deduced from the package recipe options. It requires having a look to the recipe and understanding it a bit what mingw is and how to use it, yes, but 95% of the functionality is already there, the profile is something that needs to be done just once, so even if it is not 100% automatic at the moment, not worth any further feature at the moment.

**Comment by redradist** [comment]
> 
> 
> > I do not want to have a copy of llvm toolchain, I want to have one toolchain and use it with or without conan packaging stuffs. It is much broader that a simple search of toolchain: search, manager (save, delete), create profile for it and so on
> 
> ```shell
> conan install mingw-w64/8.1@ -g virtualenv
> activate.bat
> gcc --version
> gcc (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
> ```
> 
> And done, it works. You would have the compiler environment defined, you could call `gcc` directly in the command line, no Conan involved.
> 
> > How to create profile for mingw package ?
> 
> Write a file with:
> 
> ```
> [settings]
> arch=x86_64
> os=Windows
> compiler=gcc
> compiler.version=8.1
> compiler.libcxx=libstdc++11
> compiler.threads=posix
> compiler.exception=seh
> 
> [build_requires]
> mingw-w64/8.1
> ```
> 
> This can be deduced from the package recipe options. It requires having a look to the recipe and understanding it a bit what mingw is and how to use it, yes, but 95% of the functionality is already there, the profile is something that needs to be done just once, so even if it is not 100% automatic at the moment, not worth any further feature at the moment.

@memsharded Disagree, I think it worth efforts
Lots of users do not need to now details about profile, they need to install new toolchain like @madebr suggested:
```console
conan profile new mingw --toolchain=mingw-w64/8.1@
```

All this details:
```
[settings]
arch=x86_64
os=Windows
compiler=gcc
compiler.version=8.1
compiler.libcxx=libstdc++11
compiler.threads=posix
compiler.exception=seh

[build_requires]
mingw-w64/8.1
```
are extra flexibility that probably not needed for most users

**Comment by redradist** [comment]
@memsharded I like that conan has generator `virtualenv`, I did not know about it, it simplifies building a lot !!

But I think we need one extra simplification for managing toolchains

Each toolchain could have mapped profile to it and if user would like to modify it will do it

**Comment by redradist** [comment]
@memsharded Also requiring all package to be built from the sources is just waste of resources ...
For example it is not needed to build trusted toolchains like llvm/clang, gcc linario (for embedded) and so on

**Comment by memsharded** [comment]
> I like that conan has generator virtualenv, I did not know about it, it simplifies building a lot !!

It also has many other generators, including the new ``VirtualBuildEnv`` and ``VirtualRunEnv`` ones that will be even more flexible and powerful. As I was saying, Conan has most of the features to implement this, it is mostly a matter of creating the toolchain packages.

> For example it is not needed to build trusted toolchains like llvm/clang, gcc linario (for embedded) and so on

Nobody is saying that it should always be built from sources. Conan packages can get the actual binaries from other location instead of building from sources if they don't want to. This is what the current ``mingw`` recipe is already doing, have you checked the recipe?


So again, all of this boils down to the feature of encapsulating a profile definition inside a Conan package, which is a relatively small UX advantage. We don't need that feature to have useful toolchain packages. Defining a profile takes a couple of minutes and has to be done only once in a project lifetime. Furthermore, this is a feature that would only benefit when we actually have enough toolchain packages that can and are used in practice by users, so we won't be further discussing about adding this feature until those packages are there.
